mesonbuild/linkers/linkers.py

```diff
@@ -704,8 +704,10 @@ def get_win_subsystem_args(self, value: str) -> T.List[str]:
             args = ['--subsystem,windows']
         elif 'console' in value:
             args = ['--subsystem,console']
+        elif 'efi_application' in value:
+            args = ['--subsystem,10']
```

Comment:
It looks like `objcopy` allows a subsystem of `efi-app`, but `ld` doesn't share the same capability, so we have to use the number directly.

Reply:
A code comment would make it easier to understand without having to look at this PR

Reply:
Does `ld` allow any direct numeric value?

Perhaps the correct solution would be to have a mapping table here, something like

```python
    def get_win_subsystem_args(self, value: str) -> T.List[str]:
        subsystems = {
            "windows": "windows",
            "console": "console",
            "efi_application": "10",
            # ....
        }
        newvalue = subsystems.get(value)
        if newvalue is not None:
            args = [f'--subsystem,{newvalue}']
        else:
            raise MesonException(f'Only {set(subsystems)} are supported for win_subsystem with MinGW, not "{value}".')
            ## this prints:
            ### Only {'efi_application', 'console', 'windows'} are supported for win_subsystem with MinGW, not "{value}".
            ## joining with "," except for the last item which shall be joined with the word "and" isn't conveniently done, hmm
        if ',' in value:
            args[-1] = args[-1] + ':' + value.split(',')[1]

        return self._apply_prefix(args)
```

Reply:
On second thought, this is incomplete... value -> newvalue would need to be split on "," and joined later.

Reply:
> Does `ld` allow any direct numeric value?

It appears to be based on whether `ld` is targeting `PE` or not (search for `--subsystem` in the [`ld` man page](https://www.man7.org/linux/man-pages/man1/ld.1.html)).

Reply:
> On second thought, this is incomplete... value -> newvalue would need to be split on "," and joined later.

Is this acceptable? If so, I'll commit it to this PR.

```python
def get_win_subsystem_args(self, value: str) -> T.List[str]:
        subsystems = {
            "windows": "windows",
            "console": "console",
            "efi_application": "10",
            # ....
        }
        versionsuffix = None
        if ',' in value:
            value, versionsuffix = value.split(',', 1)
        newvalue = subsystems.get(value)
        if newvalue is not None:
            if versionsuffix is not None:
                newvalue += f',{versionsuffix}'
            args = [f'--subsystem,{newvalue}']
        else:
            raise MesonException(f'Only {set(subsystems)} are supported for win_subsystem with MinGW, not "{value}".')

        return self._apply_prefix(args)
```

Reply:
That looks reasonable, yes.

Should we map some other values too?

> Numeric values are also accepted for which.

Doesn't seem to be any restriction there, my interpretation is that `ld` allows any number and simply assumes it must be a valid subsystem.

So we *could* enable all of the ones MSVC accepts.

Reply:
> That looks reasonable, yes.

Good to hear, and thank you :^). It's been committed.

> Should we map some other values too?

This sounds like something worth doing, in order to help anyone down the line making other applications. I'll look into it.

Reply:
I've looked into it: I couldn't find "concrete" documentation on what the numbers are, but I did find some [rust code](https://docs.rs/object/0.25.3/src/object/pe.rs.html#487) for PE executables that defines the subsystem constants as follows:

- `native` = 0
- `windows` = 1
- `console` = 2
- `posix` = 7
- `efi_application` = 10
- `efi_boot_service_driver` = 11
- `efi_runtime_driver` = 12
- `efi_rom` = 13
- `xbox` = 14
- `boot_application` = 16

If we are happy using this source of magic numbers, then I'll go ahead and commit + squash these changes as well.

Reply:
That rust crate documents:
> //! This module is based heavily on "winnt.h" (10.0.17763.0).

With that in mind, we can quickly find a couple other sources of magic numbers, specifically, the well-known WINE project:
https://github.com/wine-mirror/wine/blob/3ded60bd1654dc689d24a23305f4a93acce3a6f2/include/winnt.h#L2492-L2507

And the mingw-w64 sources as well:
https://github.com/mingw-w64/mingw-w64/blob/2c202976797b80703720a2bb19841d304205c936/mingw-w64-headers/include/winnt.h#L6572-L6585

These corroborate the values.

Naturally and understandably, https://docs.microsoft.com/en-us/windows/win32/api/winnt/ contains absolutely no reference to these constants at all.